streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1058,9 +1067,9 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
                 // all newly created queues should end immediately
                 _ <- newQueue.set {
                       for {
-                        id    <- nextIndex
                         queue <- Queue.bounded[Take[E1, A1]](1)
                         _     <- queue.offer(endTake)
+                        id    = UniqueKey()
```

Comment:
Inserting side-effects in pure code is notoriously difficult to reason about. In this case it should be ok (it's in the middle of a flat map and only needs non-determinism).

I don't consider this a good practice in ZIO but of course if there's a good reason for it (performance, lower code volume, etc.), then I'm all for it.

Reply:
This is not a side effect in that sense. The motivation behind UniqueKey is that we at some points want to keep track of an undefined number of fibers/iterations/etc. This was before done by arbitrarily selection int/long as the key and rolling with it. The problem with that approach is that it is every time an adhoc solution and would lead to very subtle bugs when it actually overflows.

A better key for this purposes is just an arbitrary object that does not override hashcode / equals and therefore uses its memory location as its identity. This allows us to create infinitely many distinct keys (up to the limits of ram) without any additional overhead.